File: rms-app-optimized/frontend/src/api/endpoints/careers.ts
import { api, type Envelope } from "../client";
import type { AuthUser, LoginResult } from "./auth";

export interface Job {
  rrf_id: string; job_code: string; title: string;
  department: string | null; project_name: string; location: string;
  wfh_allowed: boolean; min_experience_years: number; employment_type: string;
  needed_by_date: string | null; posted_at: string | null; openings: number;
  tags: string[]; blurb: string | null; salary_range: string | null;
}

export interface JobDetail extends Job {
  base_location: string | null;
  shift_hours: string | null;
  education_qualification: string | null;
  description: string | null;
  responsibilities: string[];
  essential_skills: string[];   // required
  desired_skills: string[];     // optional / nice-to-have
}

export interface PortalOffer {
  offer_id: string; offer_code: string; status: string; designation: string;
  ctc_annual: string; work_location: string; joining_date: string | null;
  valid_until: string | null; letter_url: string | null;
}
export interface PortalInterview {
  interview_id: string; round: string; mode: string;
  scheduled_start: string | null; scheduled_end: string | null;
  location: string | null; join_link: string | null;
}
export interface PortalApplication {
  application_id: string; rrf_id: string; job_code: string; title: string; location: string;
  current_stage: string; status: string; applied_at: string | null;
  offer: PortalOffer | null; interviews: PortalInterview[];
}
export interface PortalData {
  profile: {
    full_name: string; email: string; candidate_id: string | null;
    photo_icon_url?: string | null; photo_url?: string | null;
  };
  applications: PortalApplication[];
}

export async function listJobs(): Promise<Job[]> {
  return (await api.get<Envelope<Job[]>>("/careers/jobs")).data.data;
}

export async function getJob(jobCode: string): Promise<JobDetail> {
  return (await api.get<Envelope<JobDetail>>(`/careers/jobs/${encodeURIComponent(jobCode)}`)).data.data;
}

export async function signupCandidate(form: {
  full_name: string; email: string; password: string; phone?: string; cv: File; photo?: File | null;
}): Promise<LoginResult & { user: AuthUser }> {
  const fd = new FormData();
  fd.append("full_name", form.full_name);
  fd.append("email", form.email);
  fd.append("password", form.password);
  if (form.phone) fd.append("phone", form.phone);
  fd.append("cv_file", form.cv);
  if (form.photo) fd.append("photo", form.photo);
  return (await api.post<Envelope<LoginResult>>("/careers/signup", fd)).data.data;
}

export async function applyToJob(rrfId: string): Promise<{ application_id: string; current_stage: string; status: string }> {
  return (await api.post<Envelope<{ application_id: string; current_stage: string; status: string }>>("/careers/apply", { rrf_id: rrfId })).data.data;
}

export async function myPortal(): Promise<PortalData> {
  return (await api.get<Envelope<PortalData>>("/careers/me")).data.data;
}

export async function respondToOffer(
  offerId: string,
  action: "ACCEPT" | "DECLINE",
  comment?: string,
): Promise<{ offer_id: string; status: string; action: string }> {
  return (
    await api.post<Envelope<{ offer_id: string; status: string; action: string }>>(
      `/careers/offers/${offerId}/respond`,
      { action, comment },
    )
  ).data.data;
}

File: rms-app-optimized/backend/app/routers/careers.py
"""/careers router — public job feed + candidate self-service (signup, apply, my portal).

Public:            GET /careers/jobs, POST /careers/signup
Candidate (JWT):   POST /careers/apply, GET /careers/me
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, File, Form, UploadFile, status
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_db, require_roles
from app.models.user import User
from app.services import careers_service

router = APIRouter(prefix="/careers", tags=["careers"])


@router.get("/jobs")
async def list_jobs(db: AsyncSession = Depends(get_db)) -> dict:
    return {"success": True, "data": await careers_service.list_jobs(db)}


@router.get("/jobs/{job_code}")
async def get_job(job_code: str, db: AsyncSession = Depends(get_db)) -> dict:
    return {"success": True, "data": await careers_service.get_job(db, job_code)}


@router.post("/signup", status_code=status.HTTP_201_CREATED)
async def signup(
    full_name: str = Form(...),
    email: str = Form(...),
    password: str = Form(...),
    phone: str | None = Form(default=None),
    cv_file: UploadFile = File(...),
    photo: UploadFile | None = File(default=None),
    db: AsyncSession = Depends(get_db),
) -> dict:
    data = await cv_file.read()
    photo_data = await photo.read() if photo is not None else None
    result = await careers_service.signup(
        db, full_name=full_name, email=email, password=password, phone=phone,
        filename=cv_file.filename or "cv", content_type=cv_file.content_type or "", data=data,
        photo_data=photo_data or None,
        photo_content_type=photo.content_type if photo is not None else None,
    )
    return {"success": True, "data": result}


class ApplyBody(BaseModel):
    rrf_id: str


@router.post("/apply", status_code=status.HTTP_201_CREATED)
async def apply(
    body: ApplyBody,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_roles("CANDIDATE")),
) -> dict:
    return {"success": True, "data": await careers_service.apply(db, user, body.rrf_id)}


@router.get("/me")
async def my_portal(
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_roles("CANDIDATE")),
) -> dict:
    return {"success": True, "data": await careers_service.my_portal(db, user)}


class OfferRespondBody(BaseModel):
    action: str  # ACCEPT | DECLINE
    comment: str | None = None  # candidate's reason (optional)


@router.post("/offers/{offer_id}/respond")
async def respond_offer(
    offer_id: str,
    body: OfferRespondBody,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_roles("CANDIDATE")),
) -> dict:
    return {"success": True, "data": await careers_service.respond_offer(db, user, offer_id, body.action, body.comment)}
